### ISSUE TITLE
Object consistency

### ISSUE THREAD

**Issue by gerion0** [mesonbuild/meson#6068]
This is not an actual bug, but some ideas I got when using Meson.

While using meson, I have missed some actions that I have expected to be there by default, for example:
- The targets that are generated by generators do not have a "full_path" method.
- File objects do not have a "full_path" method.
- File objects exists only in lists.
- Only a few objects can be printed.
- Inputs are strongly typed, but this is not documented. The `extra_args` argument of `generator` for example seems to accept only strings, but not files or targets. The `input` argument of `custom_target` in contrast support strings, files, targets and lists of them.

Thinking a bit about this points, I think they can be categorized. I also try to explain what I would expect from Meson. Maybe you like this proposals. I have read most of the documentation but maybe some informations are incorrect.

# Implicit casts
Meson defines functions that have a return value (an object) and take typed input arguments. For example the input of a custom target always has to be a list of files. However, sometimes the inputs are casted implicit. In the case of the input of a custom target from a string to a list of files or from a target to a list of files. I have found that this casts are sometimes undocumented.

Another example are the print functions (`message()` and `warning()`). They accept certain arguments (`ERROR: Function accepts only strings, integers, lists and lists thereof.`) but implicitly casts them to string (but not for all meson objects).

I would expect that the cast is a property of the object not the function that take the object. So one can always call `file()` on a target object to get the output file. Or `str()` on a meson object to get a string representation, so the following is the same:
```
lst = ['a', 'b']
message(lst)
m_str = str(lst)
message(m_str)
```
The strictest way is to forbid implicit casts alltogether and accept only one input type:
```
custom_target('foo', input: flatten(my_nested_input_file_list) + file('another_input.c'), ...)
custom_target('bar', input: [file('input.c')], ...)
custom_target('error', input: 'input.c', ...) -> ERROR: argument must be a list of files
```

# Concept of super/base classes
Some objects behave very similar, e.g. all kind of targets (`custom_targets`, the targets that a `generator` creates, executables and libraries). For example all have inputs and outputs. Often they are also implicitly casted to (their output) files when used as argument. However, their common behavior is not always defined. For executables and libraries the common `build target` object exists for `custom_target` a `custom target` object and the generator targets are entirely unclear. `custom_target` and `build_target` support `full_path` while generator targets do not.

I would expect that all targets support a full_path method. In object oriented software this is realized with a common base class. Maybe this is usable as a concept for meson, too.

# Hidden state
Meson objects can have hidden internal state. For example (as far as I know) it is impossible to get the actual filepath from a file object as string. It is only possible to use it in places where sometimes also a string is used.

Other meson objects get a bunch of arguments. As far as I know it is not possible to query the values of this arguments later on (for example the inputs of a target, the actual found version of a program or the flags of an executable).

Maybe Meson can be extended in such a way, that the interal state (at least the input arguments) can always be queried.

I have seen, that a lot of functionality fixing some of these points was added in Meson with the last releases. However, for me it seems that this was often added for a specific need and not necessarily consistent. As an example when using the Python module the Python version can be queried with `language_version`. When using the Qt5 module no such option exists.

**Comment by Salamandar** [comment]
Agreed. The consistency of the "file" targets would fix a great number of issues, such as :
https://github.com/mesonbuild/meson/issues/5935, https://github.com/mesonbuild/meson/issues/5618, https://github.com/mesonbuild/meson/issues/4981, https://github.com/mesonbuild/meson/issues/4892, https://github.com/mesonbuild/meson/issues/3667, https://github.com/mesonbuild/meson/issues/3613, https://github.com/mesonbuild/meson/issues/3206, https://github.com/mesonbuild/meson/issues/3175, and a lot of others.

**Comment by Salamandar** [comment]
@jpakkane What do you think about that ?
I really think Meson needs an internal refactoring. Lots of issues are about missing methods on objects, or impossibility to install() some object types.
I'm currently browsing `interpreterbase.py` and `interpreter.py`, to find where to start.

**Comment by Salamandar** [comment]
> I would expect that the cast is a property of the object not the function that take the object. 

Yeah, to get an error like "this object has no string representation available" instead.

> So one can always call file() on a target object to get the output file. Or str() on a meson object to get a string representation, so the following is the same:

Or theobject.file() and theobject.str(). But well, that's not very different ;)

This retrieved file could be installed later (yeah, I know, not in the ideology of Meson, but sometimes that's the only way possible).

**Comment by gerion0** [comment]
Also releated: #3469

**Comment by bonzini** [comment]
Moved previous list to subissues. Leaving out a few that are more feature requests than bugs:

- [ ] https://github.com/mesonbuild/meson/issues/1550
- [ ]  https://github.com/mesonbuild/meson/issues/3206
- [ ]  https://github.com/mesonbuild/meson/issues/3469
- [ ]  https://github.com/mesonbuild/meson/issues/6675
- [ ]  https://github.com/mesonbuild/meson/issues/8295
